docs/src/userguide/freethreading.rst

```diff
@@ -160,3 +160,298 @@ to require extreme care to make it work correctly).
+  cdef object tmp
+  for i in cython.parallel.prange(10, nogil=True):
+    with gil:
+      tmp = i**2
+      total += tmp
+
+In this case, there is only a single value of ``tmp`` shared between all the threads.
+They are continuously overwriting each other's values.  Additionally, Cython does not
+currently ensure that ``tmp`` is even reference-counted in a thread-safe way,
+so you are at risk of crashes or memory-leaks in addition to getting a nonsense answer.
+
+If you do want to work with Python objects, then it is best to move them into
+a function and just have the loop call the function::
+
+  cdef int square(int x):
+    cdef object tmp = x**2
+    cdef int result = tmp
+    return result
+
+  # ...
+
+  cdef int i
+  cdef int total = 0
+  for i in cython.parallel.prange(10, nogil=True):
+    with gil:
+      total += square(i)
+
+Since ``tmp`` is now local to the function scope, each function call has its own copy
+and thus there is no conflict of Python objects between threads.
+
+Use C++ for low-level synchronization primitives
+------------------------------------------------
+
+When you must have threads interact with each other, you usually need to use
+special data types to control the access to shared data.  Python provides many of these in the
+``threading`` module.  However, sometimes it is useful to either:
+
+* avoid the Python-call overhead of the threading module,
+* use atomic variables to update numeric types in a controlled way without locking.
+
+For this our recommendation is to use the C++ standard library.  Most of these
+are available simply by "cimporting" from ``libcpp``.  In the event that Cython
+hasn't already wrapped what you want to use then you can do it yourself - our
+``libcpp`` is provided for convenience but it does nothing that can't be done
+with regular Cython code.
+
+The C standard library also provides some of these features (e.g. atomic variables
+and mutexes).  However, compiler support for the C++ standard library is better
+(in particular for MSVC) and the C++ standard library is more fully featured,
+so we recommend this first.
+
+One difficulty is with types that are not default constructable or moveable
+(e.g. ``latch``, ``semaphore``, ``barrier``).  These are difficult to
+stack-allocate because of how Cython's code-genertion works, so you
+need to heap-allocate them::
+
+  from libcpp.latch cimport latch
+
+  l = new latch(2)
+  try:
+    ...  # use the latch
+  finally:
+    del l
+
+It is also possible to use C++ to create new threads (for example, using the ``std::jthread``
+class).  This works, but we generally recommend creating threads through Python
+instead, mainly because the outermost ``with gil:`` statement in a non-Python
+created thread will not work reliably with multiple subinterpreters - this recommendation is
+therefore mainly to future-proofing your code and not restrict where it can
+be used from.  It is a fairly soft suggestion though, so feel free to ignore
```

Comment:
Maybe we should also mention that one needs to register C++ created threads with the interpreter or they cannot run Python code in them?

Reply:
That should be what the outermost ``with gil:` is for, but I'll change the text a little to clarify